src/index/hnsw/impl/IndexBruteForceWrapper.cc

```diff
@@ -0,0 +1,110 @@
+// Copyright (C) 2019-2024 Zilliz. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under the License.
+
+#include "index/hnsw/impl/IndexBruteForceWrapper.h"
+
+#include <faiss/Index.h>
+#include <faiss/MetricType.h>
+#include <faiss/cppcontrib/knowhere/impl/Bruteforce.h>
+#include <faiss/impl/AuxIndexStructures.h>
+#include <faiss/impl/DistanceComputer.h>
+#include <faiss/impl/FaissAssert.h>
+
+#include <algorithm>
+#include <memory>
+
+#include "knowhere/bitsetview.h"
+#include "knowhere/bitsetview_idselector.h"
+
+namespace knowhere {
+
+using idx_t = faiss::idx_t;
+
+// the following structure is a hack, because GCC cannot properly
+//   de-virtualize a plain BitsetViewIDSelector.
```

Comment:
the performance for selector impact may have an impact in brture force search, because it implies 90+% of filtering. In other cases, the cost of calling a virtual method is negligible compares to the cost of distance computations.

Reply:
Saying that it is also a concern.
1. IVF can have super high filtering rate but still uses virtual method.
2. To my experience the, HNSW validation is not negligible. We noticed 10% performance regression in cardinal if we try to do one more deference for that.

Reply:
Created an issue https://github.com/zilliztech/knowhere/issues/779 to track this and unblock this PR